### DISCUSSION TITLE
Help with filtering groups and results

### DISCUSSION THREAD

**Discussion by shark2k** [DirectoryTree/LdapRecord#490] [Q&A] [answered]
Hello,

I am trying to write some code that will check if a user is in a group and based on that set a variable to a specific value.
I am in an environment using Active Directory and reading the documentation I am doing the following (code from the documentation, I'm using my required values in the right places):

```php
use LdapRecord\Models\ActiveDirectory\User;

$user = User::find('cn=John Doe,ou=Users,dc=local,dc=com');

// Get all groups the user is apart of that contain 'Accounting':
$groups = $user->groups()->whereContains('cn', 'Accounting')->get();

foreach ($groups as $group) {
    if (stripos($group->getName(), 'accounting')) {
        // for testing I echoed out "In group"
        $variable = 'Accounting';
    } else {
        // for testing I echoed out "Not in group"
        $variable = 'Not accounting';
    }
}
```

Using the above code, if I use my user account as a test and I search for a group that I know I am in, I will get the correct output of "In group" on screen.  When I changed the value for whereContains so it will not find any groups, instead of getting the output of "Not in group" I just get a blank page.

Part of my problem is I'm not really good with the object oriented aspects of PHP so any OO stuff I deal with is usually built into PHP or through libraries like this, I've never created my own classes and whatnot (so I don't actually always know the proper terms).

Trying to do some troubleshooting, I tried the following to see if I could understand what was going on when a group wasn't being found (note this was put after the line from above where I did the "whereContains" portion):

```php
var_dump($groups);
```
but I get the following output (which admittedly doesn't make much sense to me):

```
object(LdapRecord\Models\Collection)#23 (2) {
  ["items":protected]=>
  array(0) {
  }
  ["escapeWhenCastingToString":protected]=>
  bool(false)
}
````

I then tried putting the following within the foreach statement to see what the value for $group was and was met with a blank screen.
```php
var_dump($group);
```

How would I go about being able to basically say, if the group exists echo "In group" (really assigning a variable but for testing purposes that's quicker) else echo "Not in group", because what I'm trying doesn't work.

Thanks for any help that can be given with this.

-Shark2k

**Comment by stevebauman** [comment, marked as answer]
Hi @shark2k,

> When I changed the value for whereContains so it will not find any groups, instead of getting the output of "Not in group" I just get a blank page.

This is because the result of the query is an empty collection, which will not be iterated over.

For example, this script will not call `die()`:

```php
$foo = [];

foreach ($foo as $bar) {
    die('This will not be reached.');
}
```

If you're looking to check whether a user has at least one group matching the query, you can instead use `exists()`:

```php
use LdapRecord\Models\ActiveDirectory\User;

$user = User::find('cn=John Doe,ou=Users,dc=local,dc=com');

if ($user->groups()->whereContains('cn', 'Accounting')->exists()) {
    // The user is apart of the "Accounting" group.
} else {
    // The user is *not* apart of the "Accounting" group.
}
```

Or, if you want to retrieve the groups matching the result:

```php
use LdapRecord\Models\ActiveDirectory\User;

$user = User::find('cn=John Doe,ou=Users,dc=local,dc=com');

$groups = $user->groups()->whereContains('cn', 'Accounting')->get();

if ($groups->isNotEmpty()) {
    // The user is apart of the "Accounting" group.
} else {
    // The user is *not* apart of the "Accounting" group.
}
```

Hope this helps! Let me know if you have further questions 👍

**Reply by shark2k** [reply]
Awesome!  Thanks so much for clarifying that as it makes sense after your explanation.
I am going to go with your second answer but the first answer is actually super helpful as well because I didn't realize you could chain the exists() method onto the query method (again not really good with OOP and have had trouble grasping the concepts).

Follow-up question to clarify/confirm my thoughts.
Going with your second answer, if I want to get the name of the group I'll still need to do a foreach on the $groups variable, correct?
```php
use LdapRecord\Models\ActiveDirectory\User;

$user = User::find('cn=John Doe,ou=Users,dc=local,dc=com');

$groups = $user->groups()->whereContains('cn', 'Accounting')->get();

if ($groups->isNotEmpty()) {
    // The user is apart of the "Accounting" group.
   foreach ($groups as $group) {
      echo $group->getName();
   }
} else {
    // The user is *not* apart of the "Accounting" group.
}
```

I'm pretty sure I will because if there are more than 1 groups that match that's the only way I am aware/familiar with to actually access the info in the collection, but just want to clarify.  I'm fine if that's the case, just want to make sure there isn't a better/more proper way to do it.

Again, thanks so much for the response and explanation.
I really do appreciate it and I really like the library that you built.

-Shark2k

**Reply by stevebauman** [reply]
Happy to help @shark2k! And thanks for your kind words, I'm glad you enjoy using LdapRecord 😄 

> Going with your second answer, if I want to get the name of the group I'll still need to do a foreach on the $groups variable, correct?

If you're looking to get the first matching group of your query, you can call the `first()` method instead of a `foreach`:

```php
$group = $user->groups()->whereContains('cn', 'Accounting')->first();

if ($group) {
    echo $group->getName();
}
```

You can also use `firstOrFail()` to retrieve the first result, or throw an exception if the user is not a member of any of the groups in the resulting query:

```php
try {
    $group = $user->groups()->whereContains('cn', 'Accounting')->firstOrFail();

    echo $group->getName();
} catch (\LdapRecord\Models\ModelNotFoundException $e) {
    // ...
}
```

**Reply by shark2k** [reply]
Great!  I'll do some tests to confirm if going with one of those options will work in my scenario and even if they won't for some reason, I appreciate you taking the time to respond and give suggestions of possible options.  It's also helpful as it's making me realize that I should be keeping in mind that methods might be able to be chained even if it's not explicitly shown in the documentation (and again that's probably more-so an outcome of me not being very good at all with OOP).

Again, really appreciate the assistance on this and also the quick response!

Thanks,
Shark2k

**Reply by stevebauman** [reply]
Ok sounds good! Happy to help -- and good luck! 😄 

Feel free to drop by with further questions 👍
